musmart/timing/swing.py

```diff
@@ -0,0 +1,141 @@
+        # Just use the first upbeat we've found
+        upbeats_bw = np.array([upbeats_bw[0]])  # keep as an array to allow iteration
+    # Iterate over all upbeats found between the two beats
+    for upbeat_bw in upbeats_bw:
+        # Compute the inter-onset intervals
+        ioi1 = upbeat_bw - beat1
+        ioi2 = beat2 - upbeat_bw
+        # Calculate the beat-upbeat ratio
+        bur = ioi2 / ioi1
+        # Express with base-2 log if required
+        if use_log2_burs:
+            bur = log2(bur)
+        yield float(bur)
+
+
+def beat_upbeat_ratio(
+        beats: Iterable,
+        upbeats: Iterable,
+        use_log2_burs: bool = False,
+        allow_multiple_burs_per_beat: bool = False
+) -> list:
+    """
+    Extracts beat-upbeat ratio (BUR) values from an array of onsets.
+
+    The beat-upbeat ratio (BUR) is introduced in [1] as a concept for analyzing the individual amount of "swing"
+    in two consecutive eighth-note beat durations. It is calculated by dividing the duration of the first
+    ("long") eighth note beat by the second ("short") beat. A BUR value of 2 indicates "perfect" swing
+    (e.g., a triplet quarter note followed by a triplet eighth note), while a BUR of 1 indicates "even"
+    eighth-note durations.
+
+    Accepts two iterables of timestamps, corresponding to `beats` and `upbeats`. These lists should be sorted,
+    but missing values can be included and will be handled internally in the function.
+
+    By default, the function assumes there is exactly one upbeat between every two consecutive beats. If there
+    are multiple upbeats between two beats, only the first upbeat is used for BUR calculation unless
+    `allow_multiple_burs_per_beat=True`, in which case a BUR is calculated for every upbeat.
+
+    Following [2], it is also common to calculate log_2 BUR values, which means that a value of 0.0 corresponds
+    to "triplet" swing. This can be enabled by setting `use_log2_burs=True`.
+
+    Parameters
+    ----------
+    beats : iterable
+        An array of beat timestamps. Should not overlap with `upbeats`.
+    upbeats : iterable
+        An array of upbeat timestamps.
+    use_log2_burs : bool, optional
+        Whether to use the log_2 of inter-onset intervals to calculate BURs, as employed in [2]. Defaults to False.
+    allow_multiple_burs_per_beat : bool, optional
+        If True, calculates a beat-upbeat ratio for every upbeat between two consecutive beats. Otherwise,
+        calculates a BUR only for the first upbeat. Defaults to False.
+
+    Returns
+    -------
+    list
+        The calculated BUR values.
+
+    Raises
+    ------
+    ValueError
+        If too few consecutive beats are provided
+
+    Examples
+    --------
+    >>> my_upbeats = np.array([0.6, 1.4])
+    >>> my_beats = np.array([0., 1., 2.])
+    >>> returned = beat_upbeat_ratio(my_beats, my_upbeats)
+    >>> [round(r, 2) for r in returned]
+    [0.67, 1.5]
```

Comment:
I'm confused about the ordering. I would have expected 1.5 to come first and 0.67 to come second? As I understand it, the first downbeat has length 0.6 and the first upbeat has length 0.4, and 0.6 / 0.4 is 1.5.

Reply:
Maybe the problem is that it depends which upbeat you compare with which downbeat? This needs to be made clearer in the docstring.